torch/csrc/autograd/functions/accumulate_grad.cpp

```diff
@@ -22,73 +22,95 @@ AccumulateGrad::AccumulateGrad(Variable variable_)
+      // the Tensor and that is fine since these are not exposed to the user.
+      update_grad_fn(new_grad_copy.detach());
     } else {
-      if (new_grad.is_sparse()) {
-        variable.grad() = new_grad.clone();
+      if (new_grad_copy.is_sparse()) {
+        update_grad_fn(new_grad_copy.clone());
       } else {
-        variable.grad() = new_grad.clone(at::MemoryFormat::Contiguous);
+        update_grad_fn(new_grad_copy.clone(at::MemoryFormat::Contiguous));
       }
     }
   } else if (!GradMode::is_enabled()) {
     // This case is not strictly necessary, but it makes the first-order only case
     // slightly more efficient.
-    Variable& grad_variable = as_variable_ref(grad);
-    if (grad_variable.is_sparse() && !new_grad.is_sparse()) {
-      // If `grad_variable` is sparse and `new_grad` is not sparse, their sum is not
-      // sparse, and we must change the TensorImpl type of `grad_variable` for it to
-      // store the result. However, changing the TensorImpl type of a tensor requires
-      // changing the tensor itself, and thus in this case we have to change the grad
-      // tensor.
-      grad_variable = new_grad + grad_variable;
+    if (variable_grad.is_sparse() && !new_grad_copy.is_sparse()) {
+      // If `grad_variable` is sparse and `new_grad_copy` is not sparse, their
+      // sum is not sparse, and we must change the TensorImpl type of
+      // `grad_variable` for it to store the result. However, changing the
+      // TensorImpl type of a tensor requires changing the tensor itself, and
+      // thus in this case we have to change the grad tensor.
+      update_grad_fn(new_grad_copy + variable_grad);
     } else {
-      // In this case we can avoid changing the grad tensor. There are three scenarios
-      // when we'll hit this case:
+      // In this case we can avoid changing the grad tensor. There are three
+      // scenarios when we'll hit this case:
       //
-      // 1. `grad_variable` is sparse, and `new_grad` is sparse.
-      // 2. `grad_variable` is dense, and `new_grad` is sparse.
-      // 3. `grad_variable` is dense, and `new_grad` is dense.
+      // 1. `grad_variable` is sparse, and `new_grad_copy` is sparse.
+      // 2. `grad_variable` is dense, and `new_grad_copy` is sparse.
+      // 3. `grad_variable` is dense, and `new_grad_copy` is dense.
       //
-      // In all of these three cases, `grad_variable += new_grad` is a valid operation
-      // which adds `new_grad` to `grad_variable` in place. `grad_variable` is thus
-      // still referring to the same tensor after the operation.
-      grad_variable += new_grad;
+      // In all of these three cases, `grad_variable += new_grad_copy` is a
+      // valid operation which adds `new_grad_copy` to `grad_variable` in place.
+      // `grad_variable` is thus still referring to the same tensor after the
+      // operation.
+      variable_grad += new_grad_copy;
```

Comment:
Why is this one not using the `update_grad_fn`? That would prevent the gradient to be recorded by the distributed autograd no?

Reply:
I'm curious why this is not detected by the tests? I did not went through all the tests in details.

Reply:
This is very surprising.
Does that mean that this function will either modify `variable_grad` inplace or call `update_grad` ?
This should be explicitly documented at least.

But it sounds like a quite weird thing to do.
Especially given that if you have get a reference to the variable_grad, you can just set it to modify the original Tensor inplace (coming either from variable.grad() or accumulatedGrads_[idx]).
That would make the API clearer no?